### ISSUE TITLE
Multi-phase drreg use hides app value from later phases

### ISSUE THREAD

**Issue by abhinav92003** [DynamoRIO/dynamorio#5121]
Today, DR supports multi-phase use for drreg, which means that clients can use drreg to get scratch regs in insertion phase and other phases as well.

#3823 added support to avoid spill slot conflicts in multi-phase use. However, different phases can still get the same scratch reg. In such "nested" spill regions, `drreg_get_app_value` won't work as expected in later phases, as it will give the value of the reg in the earlier phase, not the app value. This should be fixed, and documented until then.

Another related note:
Drreg support in non-insertion phases is bare bones -- e.g. it does not automatically restore app value before app read, or re-spill app value after app writes. This is usually fine because non-insertion phase is pretty limited. Today, we do that to expand scatter/gather instructions in app2app phase. But we should document this restriction about non-insertion phase use.

**Comment by derekbruening** [comment]
This seems like a significant problem: not the drreg_get_app_value(), which is rare, but the automated restore before app reads, which is quite common and easily corrupts and crashes the app if not done correctly every time.

Does x86 scatter/gather expansion never reserve a register that's used in its own expansion, and that's what makes this problem not occur today?  If so, is that just because the expansion only uses a small set of registers and so there are enough unused ones available, even in 32-bit?

**Comment by derekbruening** [comment]
> This seems like a significant problem: not the drreg_get_app_value(), which is rare, but the automated restore before app reads, which is quite common and easily corrupts and crashes the app if not done correctly every time.

Expanding on this: even if I reserve a reg in app2app across several instructions and I carefully restore the app value manually before app instrs that read that register, that's not enough.  The insertion phase could later use that same register, but have the wrong value stored in its slot, and think it's restoring the app value as part of its automated pass before that app read but instead it's undoing my manual restore and it's writing my tool value into that register.

**Comment by abhinav92003** [comment]
>  I carefully restore the app value manually before app instrs that read that register, that's not enough. The insertion phase could later use that same register, but have the wrong value stored in its slot

Note that the insertion phase will treat the app2app restore as an app write, and update the reg's value in its slot. So the insertion phase slot will contain the actual app value after the app2app restore.

Correct drreg usage in app2app phase requires: manually restoring the app value before app reads, and manually re-spilling app value after app writes. My understanding is that if these are done correctly, then there shouldn't be any issue. If you have a case in mind, an example would help.

**Comment by abhinav92003** [comment]
I added a new test in drreg-test to demonstrate this: PR #5124. Is this the case you were thinking of?

**Comment by derekbruening** [comment]
> > I carefully restore the app value manually before app instrs that read that register, that's not enough. The insertion phase could later use that same register, but have the wrong value stored in its slot
> 
> Note that the insertion phase will treat the app2app restore as an app write, and update the reg's value in its slot. So the insertion phase slot will contain the actual app value after the app2app restore.

This would only happen if drreg treated both app and tool instructions from app2app the same (as app instructions): yes, this does seem to be the case.

But given that: there is no new issue raised with treating an annotation label as an app read, and so no concern with Proposal A from https://github.com/DynamoRIO/dynamorio/issues/5118#issuecomment-925425107, unlike in the original discussion.

**Comment by abhinav92003** [comment]
> whoever is doing restores before app reads in app2app needs to do the same before an annotation label is all.

Yes, something like PR #5126. Thoughts on submitting that PR so that we can point to that example for how app2app and insertion phases need to work together?
